clang/test/Sema/ext_vector_casts.c

```diff
@@ -116,7 +116,7 @@ static void splats(int i, long l, __uint128_t t, float f, double d) {
   
   vf = 1 + vf;
   vf = l + vf; // expected-warning {{implicit conversion from 'long' to 'float2' (vector of 2 'float' values) may lose precision}}
-  vf = 2.0 + vf;
+  vf = 2.0 + vf; // expected-warning {{implicit conversion between floating point types of different sizes}}
```

Comment:
`vd` is a vector of 2 `double` values. Since this is a conversion from `float`  to  `double` there is no precision loss and a warning is not triggered here. Also,  if the vector cast is cast between two vectors of the same size, it is a bitcast, not a conversion - https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaChecking.cpp#L13796

Reply:
> vd is a vector of 2 double values. Since this is a conversion from float to double there is no precision loss and a warning is not triggered here.

The new diagnostic is about size conversions, not precision changes.

> Also, if the vector cast is cast between two vectors of the same size, it is a bitcast, not a conversion - https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaChecking.cpp#L13796

If we're diagnosing this as being a size conversion:
```
  vf = 2.0 + vf; // expected-warning {{implicit conversion between floating point types of different sizes}}
```
what is your rationale for not diagnosing this similarly:
```
vd = 2.0f + vd;
```

Reply:
After debugging, it turns out that this is a case of double promotion. 
If we pass the flag `-Wdouble-promotion`, we will get the following warning:

`implicit conversion increases floating-point precision: 'float' to 'double2' (vector of 2 'double' values)`

Reply:
Oh god. I had no idea we had a diagnostic like that! UGH. So it sounds like there's yet another diagnostic in this same general space that we need to think about (it's a GCC diagnostic Clang added for compatibility, as best I can tell). This covers some of what the user was asking for, but not all of it, because it only handles an increase in precision not a change in size. 

(I *hate* the diagnostic group it's under because this is *not* a promotion (default argument promotions are what happen when you call a variadic function and pass a float, as in: `void func(int a, ...); void foo(void) { func(1, 1.0f); }` the `1.0f` is promoted to a double in that case).) 

I think we need to think about how these diagnostics interact a bit more carefully. It's not clear to me which diagnostic to prefer in what circumstance yet, and there's significant overlap between the size conversion warning you're working on and the existing `-Wdouble-promotion` warning, it seems.

Reply:
> It's not clear to me which diagnostic to prefer in what circumstance yet


The double promotion warning is [emitted](https://github.com/intel/llvm/blob/78445c3b6ed0febdac24445928fecf40a471f01f/clang/lib/Sema/SemaChecking.cpp#L13851) when the difference in [floating rank](https://clang.llvm.org/doxygen/ASTContext_8cpp.html#a9d365d6aac4fa797c59739c573ff860ba771f23e839155dcc6e928f1dda070472) is negative. i.e when there is a conversion from `float` to `double`, `double` to `long double` , `float` to `float128`
when the [order](https://github.com/intel/llvm/blob/78445c3b6ed0febdac24445928fecf40a471f01f/clang/lib/Sema/SemaChecking.cpp#L13847) is less than 0, it captures both size as well as precision changes.
Maybe we can add a  [IsSameFloatAfterCast](https://github.com/intel/llvm/blob/78445c3b6ed0febdac24445928fecf40a471f01f/clang/lib/Sema/SemaChecking.cpp#L12557) check when order is < 0 and emit a new size warning?

Reply:
Also, the double promotion warning is only available on host code currently. Since the user's original ask was for floating size changes in device code, would the changes in the PR be useful for that? Just a thought :)

Reply:
Ugh, so we're *so close* to having everything we need with the existing warnings. We have a loses precision warning, and we have a gains precision warning, and all we're missing is a "precision didn't change but size did" warning. But in terms of warning priority, I think we should prefer the loses precision warning over the gains precision warning, over the changes size warning. The first case is the most important because it can introduce errors. The second case is more important than the third because gaining extra precision may also introduce errors. The third case is the fallback for "please just tell me when there's a conversion".

That said, the testing matrix for this will be a hoot -- we want to test all the combinations of these diagnostics being enabled/disabled to make sure we don't accidentally drop or duplicate diagnostics.

Reply:
Was this test added as requested above:
`vd = 2.0f + vd;`

Reply:
This test case has been included in `clang/test/Sema/precision-and-size-conversion.c`

Reply:
> This test case has been included in `clang/test/Sema/precision-and-size-conversion.c`

Sorry, I missed this test, thanks.   Does the diagnostic get issued on the device?   Did that get addressed?

Reply:
The double promotion diagnostic is only available on host. Enabling it on device wasn't part of the scope of this change set.

Reply:
I understand that that wasn't part of the original ask of this PR, but given what we discovered along the way, I feel it completes what might be useful for the user on device.  @AaronBallman, what do you feel about this?

Can you explain what you mean by "only available on host"?  Is the codepath somehow disabled on target?  Thanks.